File: 剑指offer/49.丑数.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        dp = [0] * n
        dp[0] = 1

        a = 0
        b = 0
        c = 0

        for i in range(2, n):
            n1 = dp[a] * 2
            n2 = dp[b] * 3
            n3 = dp[c] * 5

            dp[i] = min(n1, n2, n3)

            if dp[i] == n1:
                a += 1
            if dp[i] == n2:
                b += 1
            if dp[i] == n3:
                c += 1

        return dp[-1]
        

File: 剑指offer/49.丑数.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int nthUglyNumber(int n)
    {
        vector<int> dp(n + 1);
        dp[1] = 1;

        int p2 = 1;
        int p3 = 1;
        int p5 = 1;

        for (int i = 2; i <= n; i++) {
            int n2 = dp[p2] * 2;
            int n3 = dp[p3] * 3;
            int n5 = dp[p5] * 5;

            dp[i] = min(min(n2, n3), n5);

            if (dp[i] == n2) {
                p2 += 1;
            }
            if (dp[i] == n3) {
                p3 += 1;
            }
            if (dp[i] == n5) {
                p5 += 1;
            }
        }

        return dp[n];
    }
};